Compilation failure with C++17

I’m using Microsoft (R) C/C++ Optimizing Compiler Version 19.24.28319 for x64.
I’m compiling ocaml-mccs 1.1+11.

The incriminated line seems to be:
https://github.com/AltGr/ocaml-mccs/blob/93cd3b25878c1a2bde5cab0a5c85a376936dc183/src/mccs_stubs.cpp#L316

Here’s the report:
```
(cd _build/default/src_ext/mccs/src && "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\bin\HostX64\x64\cl.exe" -I C:/cygwin64/opt/ocaml-platform/lib/ocaml -I ../../cudf -I ../../extlib/src -I glpk -nologo -O2 -Gy- -MD -I
. /EHsc -DUSEGLPK /Fomccs_stubs.obj -c mccs_stubs.cpp)
mccs_stubs.cpp
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\xtree(1708): error C3848: expression having type 'const CUDFPackage_comparator' would lose some const-volatile qualifiers in order to call 'bool CUDFPackage_comparator::operat
or ()(CUDFVersionedPackage *,CUDFVersionedPackage *)'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\xtree(1005): note: see reference to function template instantiation 'std::_Tree_find_result<std::_Tree_node<CUDFVersionedPackage *,std::_Default_allocator_traits<_Alloc>::void
_pointer> *> std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>>::_Find_lower_bound<_Key>(const _Keyty &) const' being compiled
        with
        [
            _Alloc=std::allocator<CUDFVersionedPackage *>,
            _Kty=CUDFVersionedPackage *,
            _Pr=CUDFPackage_comparator,
            _Key=CUDFVersionedPackage *,
            _Keyty=CUDFVersionedPackage *
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\xtree(1005): note: see reference to function template instantiation 'std::_Tree_find_result<std::_Tree_node<CUDFVersionedPackage *,std::_Default_allocator_traits<_Alloc>::void
_pointer> *> std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>>::_Find_lower_bound<_Key>(const _Keyty &) const' being compiled
        with
        [
            _Alloc=std::allocator<CUDFVersionedPackage *>,
            _Kty=CUDFVersionedPackage *,
            _Pr=CUDFPackage_comparator,
            _Key=CUDFVersionedPackage *,
            _Keyty=CUDFVersionedPackage *
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\xtree(1331): note: see reference to function template instantiation 'std::pair<std::_Tree_node<CUDFVersionedPackage *,std::_Default_allocator_traits<_Alloc>::void_pointer> *,b
ool> std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>>::_Emplace<CUDFVersionedPackage*const &>(CUDFVersionedPackage *const &)' being compiled
        with
        [
            _Alloc=std::allocator<CUDFVersionedPackage *>,
            _Kty=CUDFVersionedPackage *,
            _Pr=CUDFPackage_comparator
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\xtree(1331): note: see reference to function template instantiation 'std::pair<std::_Tree_node<CUDFVersionedPackage *,std::_Default_allocator_traits<_Alloc>::void_pointer> *,b
ool> std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>>::_Emplace<CUDFVersionedPackage*const &>(CUDFVersionedPackage *const &)' being compiled
        with
        [
            _Alloc=std::allocator<CUDFVersionedPackage *>,
            _Kty=CUDFVersionedPackage *,
            _Pr=CUDFPackage_comparator
        ]
mccs_stubs.cpp(316): note: see reference to function template instantiation 'std::pair<std::_Tree_const_iterator<std::_Tree_val<std::_Tree_simple_types<CUDFVersionedPackage *>>>,bool> std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>>::insert<false,0>(CUDFVersionedPack
age *const &)' being compiled
        with
        [
            _Kty=CUDFVersionedPackage *,
            _Pr=CUDFPackage_comparator,
            _Alloc=std::allocator<CUDFVersionedPackage *>
        ]
mccs_stubs.cpp(316): note: see reference to function template instantiation 'std::pair<std::_Tree_const_iterator<std::_Tree_val<std::_Tree_simple_types<CUDFVersionedPackage *>>>,bool> std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>>::insert<false,0>(CUDFVersionedPack
age *const &)' being compiled
        with
        [
            _Kty=CUDFVersionedPackage *,
            _Pr=CUDFPackage_comparator,
            _Alloc=std::allocator<CUDFVersionedPackage *>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\xtree(1698): error C3848: expression having type 'const CUDFPackage_comparator' would lose some const-volatile qualifiers in order to call 'bool CUDFPackage_comparator::operat
or ()(CUDFVersionedPackage *,CUDFVersionedPackage *)'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\xtree(1006): note: see reference to function template instantiation 'bool std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>>::_Lower_bound_duplicate<_Key>(std::_Tree_node<CU
DFVersionedPackage *,std::_Default_allocator_traits<_Alloc>::void_pointer> *const ,const _Keyty &) const' being compiled
        with
        [
            _Kty=CUDFVersionedPackage *,
            _Pr=CUDFPackage_comparator,
            _Alloc=std::allocator<CUDFVersionedPackage *>,
            _Key=CUDFVersionedPackage *,
            _Keyty=CUDFVersionedPackage *
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.24.28314\include\xtree(1006): note: see reference to function template instantiation 'bool std::_Tree<std::_Tset_traits<_Kty,_Pr,_Alloc,false>>::_Lower_bound_duplicate<_Key>(std::_Tree_node<CU
DFVersionedPackage *,std::_Default_allocator_traits<_Alloc>::void_pointer> *const ,const _Keyty &) const' being compiled
        with
        [
            _Kty=CUDFVersionedPackage *,
            _Pr=CUDFPackage_comparator,
            _Alloc=std::allocator<CUDFVersionedPackage *>,
            _Key=CUDFVersionedPackage *,
            _Keyty=CUDFVersionedPackage *
        ]
```

G++ in C++14 compiles ocaml-mccs just fine. It’s weird, because a sample program I tested on MSVC reported for [`_MSVC_LANG`](https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=vs-2019) `201402`.

G++ in C++17 mode also fails to compile ocaml-mccs, for what seems to be the same cause.

```
Command [101] exited with code 1:
$ (cd _build/default/src && /usr/bin/gcc -I /usr/lib/ocaml -I /home/antonin/.opam/default/lib/bytes -I /home/antonin/.opam/default/lib/cudf -I /home/antonin/.opam/default/lib/extlib -I glpk -O2 -fno-strict-aliasing -fwrapv -fPIC -I . -std=c++17 -DUSEGLPK -g -o mccs_stubs.o -c mccs_stubs.cpp)
In file included from /usr/include/c++/9.2.1/map:60,
                 from mccs_stubs.cpp:14:
/usr/include/c++/9.2.1/bits/stl_tree.h: In instantiation of ‘static const _Key& std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::_S_key(std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::_Const_Link_type) [with _Key = CUDFVersionedPackage*; _Val = CUDFVersionedPackage*; _KeyOfValue = std::_Identity<CUDFVersionedPackage*>; _Compare = CUDFPackage_comparator; _Alloc = std::allocator<CUDFVersionedPackage*>; std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::_Const_Link_type = const std::_Rb_tree_node<CUDFVersionedPackage*>*]’:
/usr/include/c++/9.2.1/bits/stl_tree.h:2095:47:   required from ‘std::pair<std::_Rb_tree_node_base*, std::_Rb_tree_node_base*> std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::_M_get_insert_unique_pos(const key_type&) [with _Key = CUDFVersionedPackage*; _Val = CUDFVersionedPackage*; _KeyOfValue = std::_Identity<CUDFVersionedPackage*>; _Compare = CUDFPackage_comparator; _Alloc = std::allocator<CUDFVersionedPackage*>; std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::key_type = CUDFVersionedPackage*]’
/usr/include/c++/9.2.1/bits/stl_tree.h:2148:4:   required from ‘std::pair<std::_Rb_tree_iterator<_Val>, bool> std::_Rb_tree<_Key, _Val, _KeyOfValue, _Compare, _Alloc>::_M_insert_unique(_Arg&&) [with _Arg = CUDFVersionedPackage* const&; _Key = CUDFVersionedPackage*; _Val = CUDFVersionedPackage*; _KeyOfValue = std::_Identity<CUDFVersionedPackage*>; _Compare = CUDFPackage_comparator; _Alloc = std::allocator<CUDFVersionedPackage*>]’
/usr/include/c++/9.2.1/bits/stl_set.h:511:48:   required from ‘std::pair<typename std::_Rb_tree<_Key, _Key, std::_Identity<_Tp>, _Compare, typename __gnu_cxx::__alloc_traits<_Alloc>::rebind<_Key>::other>::const_iterator, bool> std::set<_Key, _Compare, _Alloc>::insert(const value_type&) [with _Key = CUDFVersionedPackage*; _Compare = CUDFPackage_comparator; _Alloc = std::allocator<CUDFVersionedPackage*>; typename std::_Rb_tree<_Key, _Key, std::_Identity<_Tp>, _Compare, typename __gnu_cxx::__alloc_traits<_Alloc>::rebind<_Key>::other>::const_iterator = std::_Rb_tree_const_iterator<CUDFVersionedPackage*>; std::set<_Key, _Compare, _Alloc>::value_type = CUDFVersionedPackage*]’
mccs_stubs.cpp:316:43:   required from here
/usr/include/c++/9.2.1/bits/stl_tree.h:780:8: error: static assertion failed: comparison object must be invocable as const
  780 |        is_invocable_v<const _Compare&, const _Key&, const _Key&>,
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

--- a/src/cudf.h
+++ b/src/cudf.h
@@ -118,7 +118,7 @@ public:
 // Compares two versioned package (used by ordered set packages likes CUDFVersionedPackageSet)
 class CUDFPackage_comparator {
 public:
-  bool operator()(CUDFVersionedPackage *p1, CUDFVersionedPackage *p2) {
+  bool operator()(CUDFVersionedPackage *p1, CUDFVersionedPackage *p2) const {
     if (p1->version < p2->version)
       return true;
     else

